numba/np/linalg.py

```diff
@@ -2792,9 +2792,10 @@ def kron_impl(a, b):
     fix_b = _kron_normaliser_impl(b)
     ret_c = _kron_return(a, b)
 
-    # this is fine because the ufunc for the Hadamard product
-    # will reject differing dtypes in a and b.
     dt = getattr(a, 'dtype', a)
+    if (isinstance(a, types.Array) and isinstance(b, types.Array) and
+       a.dtype != b.dtype):
+        raise NumbaTypeError("Both types must be the same in numpy.kron.")
```

Comment:
I'm not sure that this check is currently sufficient. Consider:
```python
from numba import njit
import numpy as np

@njit
def foo(x, y):
    return np.kron(x, y)

args = (1, np.array([1.2, 3.4]))
print(foo(*args))
print(foo.py_func(*args))
```
which exhibits the same problem as in #8586. Perhaps something like this might work?

```suggestion
    dta = getattr(a, 'dtype', a)
    dtb = getattr(b, 'dtype', b)
    if dta != dtb:
        raise NumbaTypeError(("All the arguments must have the same "
                              f"type/dtype. Have: '{dta}' and '{dtb}'."))
    dt = dta
```

Reply:
I initially did that, but it breaks all the unit tests.